core/services/pipeline/runner.go

```diff
@@ -22,6 +22,7 @@ type (
 		CreateRun(ctx context.Context, jobID int32, meta map[string]interface{}) (int64, error)
 		AwaitRun(ctx context.Context, runID int64) error
 		ResultsForRun(ctx context.Context, runID int64) ([]Result, error)
+		UpsertErrorFor(specID int32, description string)
```

Comment:
This name reveals the implementation, something like "RecordError" might be better.